reframe/listener.go

```diff
@@ -0,0 +1,507 @@
+			// if cid hasn't expired - adding it to the replacement chunk
+			if _, ok := cidsToRemove[c]; !ok {
+				newChunk.Cids[c] = struct{}{}
+				continue
+			}
+
+			// cleaning up the expired cid
+			err := listener.dsWrapper.deleteCidTimestamp(ctx, c)
+			if err != nil {
+				log.Errorw("Error cleaning up timestamp by cid index.", "err", err)
+			}
+			listener.nodesLl.Remove(cidsToRemove[c])
+			delete(listener.listNodeByCid, c)
+			delete(cidsToRemove, c)
+			listener.stats.cidsExpired++
+		}
+		// only generating a new chunk if it has some cids left in it
+		if len(newChunk.Cids) > 0 {
+			newChunk.ContextID = generateContextID(newChunk.Cids, listener.nonceGen())
+			newCtxIdStr := contextIDToStr(newChunk.ContextID)
+
+			log.Infow("Generating a replacement chunk.", "new", newCtxIdStr, "old", oldCtxIdStr, "size", len(newChunk.Cids))
+
+			err = listener.notifyPutAndPersist(ctx, newChunk)
+			if err != nil {
+				return err
+			}
+		} else {
+			log.Infof("No CIDs left to generate a replacement chunk for %s.", contextIDToStr(chunkToRemove.ContextID))
+		}
+	}
+
+	// we might have still some expired cids left, that didn't have any chunk associated to them
+	for k, v := range cidsToRemove {
+		// cleaning up the expired cid
+		err := listener.dsWrapper.deleteCidTimestamp(ctx, k)
+		if err != nil {
+			log.Errorw("Error cleaning up timestamp by cid index.", "err", err)
+		}
+		listener.nodesLl.Remove(v)
+		delete(listener.listNodeByCid, k)
+	}
+
+	return nil
+}
+
+func (listener *ReframeListener) notifyRemoveAndPersist(ctx context.Context, chunk *cidsChunk) error {
+	ctxIdStr := contextIDToStr(chunk.ContextID)
+	log.Infof("Notifying Remove for chunk=%s", ctxIdStr)
+
+	// notifying the indexer
+	_, err := listener.engine.NotifyRemove(ctx, listener.provider(), chunk.ContextID)
+	if err != nil && err != provider.ErrAlreadyAdvertised {
+		return err
+	}
+	listener.stats.removeAdsSent++
+
+	// cleaning up the datastore
+	err = listener.dsWrapper.deleteChunkByContextID(ctx, chunk.ContextID)
+	if err != nil {
+		return err
+	}
+
+	// cleaning up the in-memory indexes
+	delete(listener.chunkByContextId, ctxIdStr)
+	for c := range chunk.Cids {
+		delete(listener.chunkByCid, c)
+	}
+
+	return nil
+}
+
+func (listener *ReframeListener) notifyPutAndPersist(ctx context.Context, chunk *cidsChunk) error {
+	ctxIdStr := contextIDToStr(chunk.ContextID)
+	log.Infof("Notifying Put for chunk=%s, provider=%s, addrs=%q, cidsTotal=%d", ctxIdStr, listener.provider(), listener.addrs(), len(chunk.Cids))
+
+	// we need to notify put first and only if it succeeds record the data to the datastore
+	// because the datastore contains only chunks that have already been successfully advertised
+	_, err := listener.engine.NotifyPut(ctx, &peer.AddrInfo{ID: listener.provider(), Addrs: listener.addrs()}, chunk.ContextID, bitswapMetadata)
+	if err != nil && err != provider.ErrAlreadyAdvertised {
+		log.Errorf("Error NotifyingPut for chunk %s", chunk.ContextID)
```

Comment:
Can you just return the error and log it once higher in the call stack?

Reply:
Yep, removed